server/src/main/java/org/elasticsearch/action/support/broadcast/unpromotable/TransportBroadcastUnpromotableAction.java

```diff
@@ -82,6 +90,30 @@ protected void doExecute(Task task, Request request, ActionListener<ActionRespon
         }
     }
 
+    private void failShard(ShardRouting shardRouting, ClusterState clusterState, ActionListener<Object> l, Exception e) {
+        shardStateAction.remoteShardFailed(
+            shardRouting.shardId(),
+            shardRouting.allocationId().getId(),
+            clusterState.metadata().index(shardRouting.getIndexName()).primaryTerm(shardRouting.shardId().getId()),
+            true,
+            "mark unpromotable copy as stale after refresh failure",
+            e,
+            new ActionListener<>() {
+                @Override
+                public void onResponse(Void unused) {
+                    // TODO Only fail if were unable to mark the shard as stale?
```

Comment:
Yes. If we successfully fail the shard, no need to fully fail the broadcast request. We might add an debug log message though.